File: Haskell/arrow/move_knight/move_knight_many.hs
import Control.Arrow

type KnightPos = (Int, Int)

moveKnight :: KnightPos -> [KnightPos]
moveKnight (c, r) = filter onBoard
	[
		(c + 2, r - 1), (c + 2, r + 1),
		(c - 2, r - 1), (c - 2, r + 1),
		(c + 1, r - 2), (c + 1, r + 2),
		(c - 1, r - 2), (c - 1, r + 2)
	]
	where onBoard (c', r') = c' `elem` [1..8] && r' `elem` [1..8]

inMany :: Int -> Kleisli [] KnightPos KnightPos
inMany x = foldr (>>>) returnA (replicate x (Kleisli moveKnight))

canReachInMany :: Int -> KnightPos -> KnightPos -> Bool
canReachInMany x end = runKleisli (inMany x) >>> elem end

main = do
	print $ runKleisli (inMany 3) $ (6, 2)

	print $ canReachInMany 3 (6, 1) $ (6, 2)
	print $ canReachInMany 3 (7, 3) $ (6, 2)

File: blog/20140810/kleisli_arrow/scalaz/src/main/scala/MoveKnightMany.scala
package sample

import scalaz._
import Scalaz._

object MoveKnightMany extends App {

	type KnightPos = Tuple2[Int, Int]

	val inRange = (p: Int) => 1 to 8 contains p

	val moveKnight = (p: KnightPos) => List(
		(p._1 + 2, p._2 - 1), (p._1 + 2, p._2 + 1),
		(p._1 - 2, p._2 - 1), (p._1 - 2, p._2 + 1),
		(p._1 + 1, p._2 - 2), (p._1 + 1, p._2 + 2),
		(p._1 - 1, p._2 - 2), (p._1 - 1, p._2 + 2)
	).filter { case (x, y) => inRange(x) && inRange(y) }

	val inMany = (x: Int) => List.fill(x) { Kleisli(moveKnight) }.reduce { (a, b) => a >>> b }
	// 以下でも可
	// val inMany = (x: Int) => List.fill(x) { Kleisli(moveKnight) }.reduce { (a, b) => a >=> b }

	val canReachInMany = (x: Int) => (end: KnightPos) => inMany(x).run >>> { xs => xs.contains(end) }

	(6, 2) |> inMany(3) |> println

	(6, 2) |> canReachInMany(3)(6, 1) |> println
	(6, 2) |> canReachInMany(3)(7, 3) |> println
}
